### ISSUE TITLE
Let world coordinates wrap at an arbitrary power of 2

### ISSUE THREAD

**Issue by byorgey** [swarm-game/swarm#606]
This issue is still open for the idea of being able to specify the power of two at which world coordinates wrap.  *i.e.* add a field to the `world` in a scenario `.yaml` file like `size: n` which has a natural number specifying the power of two, i.e. `size: 32` would be the default, but you could also specify smaller values.  There can also be a Boolean field `wrap` which specifies whether the world should wrap around or just stop at the edges.

Below is the original issue description for switching world coordinates to `Int32`, which has now been done.

----

Right now we use `Int64` for world coordinates.  However, this is a little ridiculous:
- Even travelling at 1024 units per second, it would take a robot 285 million years of real-world time to travel all the way to the "edge" of the world (or rather, the place where the world wraps around).
- If each cell represents 1 square meter, this would be a planet about 42 billion times the size of Jupiter. 

On the other hand, with `Int32` for coordinates instead:
- It would take a robot travelling at 1024 cells/s about 24 days to reach the edge of the world.
- At 1m / cell, it would represent a planet about 9 times the size of Jupiter.

This is obviously still plenty big enough that no one will ever run out of space, but seems more reasonable/realistic.  ~~And if we say that each cell is 100m x 100m (not that it really matters), the planet would be about the right size (and would mean that features like forests and lakes etc. are kind of the right size too).~~ (Oops, ignore that last comment, I went the wrong way --- it should be 1 cm / cell, in which case lakes and forests are *definitely* not the right size. :laughing: )

**Comment by xsebek** [comment]
Could we add some (`new`)`type Point`? I am not sure if we use `V2` for anything else.

**Comment by byorgey** [comment]
Yes, making a `newtype Loc = Loc (V2 Int32)` is a good idea.  I don't think we use `V2` for anything else but just on principle.

**Comment by xsebek** [comment]
Generalising this idea a bit, we could make the world coordinates "wrap around" an arbitrary number.

For example, we could have a small world that is only 128x128 cells large. Not just adding a wall, but an actual finite world.

The funny thing would be that it is a torus, so if you go east, you will come back from the west, arcade style! :space_invader:

It would be cool to create the city Sigil from [Planescape: Torment](https://en.wikipedia.org/wiki/Planescape:_Torment). :smile: The city is inside the torus above the insanely high mountain:
![Sigil](https://upload.wikimedia.org/wikipedia/en/0/02/Pgp_007_01.jpg)

**Comment by byorgey** [comment]
Ah, nice idea!  So this could be an additional field in a scenario description.  We might want to limit it to powers of 2 for efficiency - so we only have to do a bitwise AND with a mask instead of doing an arbitrary mod operation.  I imagine it could make a big difference since we do lots of coordinate calculations.

**Comment by byorgey** [comment]
I'm currently imagining there can be two fields: 
- `size: n` specifies that the world should be $2^n \times 2^n$
- `wrap` is a boolean specifying whether the world should wrap around at the edges.  When `true`, the world is a torus; when `false`, the world just stops at the edge, and you can't move past the edge as if there were an unwalkable entity in the cell just beyond it.
